contracts/dao/neutron-flashloans/src/contract.rs

```diff
@@ -0,0 +1,403 @@
+/// a submessage with a `reply_on_success` strategy, meaning if it fails, the transaction is
+/// reverted.
+pub fn execute_request_loan(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    amount: Vec<Coin>,
+) -> Result<Response<NeutronMsg>, ContractError> {
+    let config: Config = CONFIG.load(deps.storage)?;
+
+    // Reentrancy guard
+    if ACTIVE_LOAN.load(deps.storage)?.is_some() {
+        return Err(FlashloanAlreadyActive {});
+    }
+
+    // Check that the amount does not have duplicate or zero coins
+    validate_amount(amount.clone())?;
+
+    // Getting the current balances of the source is necessary to make sure that the loan
+    // was returned, and the fee was paid.
+    let pre_loan_balances = get_pre_loan_balances(&deps, config.source.clone(), amount.clone())?;
+
+    // Calculate the fee to be paid by the borrower for receiving the loan
+    let fee = calculate_fee(amount.clone(), config.fee_rate)?;
+
+    // Calculate the expected balance of the source after the loan has been returned
+    // and the fee has been paid as (requested amount + fee).
+    let expected_balances = calculate_expected_balances(pre_loan_balances, fee.clone())?;
+
+    // Save all the information necessary to continue processing the loan request in the reply()
+    // handler.
+    ACTIVE_LOAN.save(
+        deps.storage,
+        &Some(ActiveLoan {
+            borrower: info.sender.clone(),
+            amount: amount.clone(),
+            fee,
+            expected_balances,
+        }),
+    )?;
+
+    // Send a (stargate -> authz -> bank) /cosmos.bank.v1beta1.MsgSend submessage with
+    // reply_on_success strategy (we want the transaction to be simply reverted in case of an
+    // error).
+    let msg_send = get_stargate_authz_bank_send_msg(env, config, info.sender, amount);
+    Ok(Response::new()
+        .add_submessage(SubMsg::reply_on_success(msg_send, AUTHZ_BANK_SEND_REPLY_ID))
+        .add_attribute("action", "execute_get_loan"))
+}
+
+// Check that the amount does not have duplicate or zero coins
+fn validate_amount(amount: Vec<Coin>) -> Result<(), ContractError> {
+    let mut denoms: HashSet<String> = HashSet::new();
+    for coin in amount {
+        if coin.amount.eq(&Uint128::zero()) {
+            return Err(ContractError::ZeroRequested { denom: coin.denom });
+        }
+
+        if denoms.contains(&coin.denom) {
+            return Err(ContractError::DuplicateDenoms {});
+        }
+
+        denoms.insert(coin.denom);
+    }
+
+    Ok(())
+}
+
+/// Returns the list of current balances on the contract's account for the coins that were
+/// requested by the borrower. If any actual coin balance is lower than the requested amount,
+/// returns an error.
+fn get_pre_loan_balances(
+    deps: &DepsMut,
+    source: Addr,
+    requested_amount: Vec<Coin>,
+) -> Result<Vec<Coin>, ContractError> {
+    // Prepare the query
+    let all_balances_query = BankQuery::AllBalances {
```

Comment:
I don't think this is safe.

I can spam the `source` address with an _infinite_ amount of TF tokens, so after the certain number of created tokens your `AllBalances` query will return a gas error, thus users can't get new flashloans

Reply:
Do you suggest having an allow list of tokens that we are ready to loan?

Reply:
Either that, or do a separate `BankQuery::Balance { address: String, denom: String }` query.
That way query can be broken only if a user himself provides a lot of tokens to loan. But it's kinda his problem, not contract's

Reply:
Makes sense, I like the separate query approach more